### PR TITLE
Transaction service refactoring

### PR DESCRIPTION
**Pull request by marcinwadon**

### CODE REVIEW

File: src/main/scala/org/constellation/primitives/storage/Storage.scala [hunk 1]
```diff
@@ -32,4 +32,8 @@ trait Storage[F[_], K, V] {
   def toMap(): F[Map[K, V]]
 
   def cacheSize(): Long
+
+  def getLast20: F[List[V]]
+
+  def getLast20Sync: List[V]
```
Comment (Nikolaj-K) [inline src/main/scala/org/constellation/primitives/storage/Storage.scala]: Can `getLast` and its variants be a function and parametrized by the same value below, i.e.

```scala
private val maxQueueSize = 20

Comment (marcinwadon) [reply src/main/scala/org/constellation/primitives/storage/Storage.scala]: it needs more refactoring and dynamic queue and if such feature will be needed, we'll do it separately

File: src/main/scala/org/constellation/consensus/EdgeProcessor.scala [hunk 2]
```diff
@@ -469,19 +470,23 @@ object Snapshot {
...
-//          .flatMap(_ => dao.acceptedTransactionService.remove(Set(tx.hash)))
-          .flatMap(_ => IO { println(s"------- Should remove tx: ${tx.hash}") })
-          .flatTap(_ => dao.metrics.incrementMetricAsync("snapshotAppliedBalance"))
-          .unsafeRunSync()
-      }
+//      CheckpointService.fetchTransactions(cb.transactionsMerkleRoot)
+//        .toList
+//        .map(tx => dao.addressService.transferSnapshot(tx)
+//          .flatTap(_ => dao.metrics.incrementMetricAsync("snapshotAppliedBalance")))
+//        .sequence
```
Comment (tylerprete) [inline src/main/scala/org/constellation/consensus/EdgeProcessor.scala]: keeping this around as comments for a reason?

File: src/main/scala/org/constellation/consensus/RoundManager.scala [hunk 3]
```diff
@@ -37,6 +37,7 @@ class RoundManager(roundTimeout: FiniteDuration)(implicit dao: DAO)
     case StartNewBlockCreationRound if !ownRoundInProgress =>
       ownRoundInProgress = true
       createRoundData(dao).fold {
+        println("------------------ ERROR create round data")
```
Comment (tylerprete) [inline src/main/scala/org/constellation/consensus/RoundManager.scala]: you want this to stay in? Any reason not to use logger?

File: src/main/scala/org/constellation/consensus/EdgeProcessor.scala (hunk 2 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/consensus/EdgeProcessor.scala]: not really, haven't pushed cleaned up version, sorry

File: src/main/scala/org/constellation/consensus/RoundManager.scala (hunk 3 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/consensus/RoundManager.scala]: it was for debugging only, cleaned up right now.

File: src/main/scala/org/constellation/util/Metrics.scala [hunk 4]
```diff
@@ -228,7 +228,7 @@ class Metrics(periodSeconds: Int = 1)(implicit dao: DAO)
       updateMetric("channelCount", dao.threadSafeMessageMemPool.activeChannels.size)
 
       // Get TransactionServiceDBSize
-      val txServiceMap = dao.transactionService.getMetricsMap
+      val txServiceMap = dao.transactionService.getMetricsMap.unsafeRunSync()
```
Comment (wkoszycki) [inline src/main/scala/org/constellation/util/Metrics.scala]: you could use updateMetricAsync and chain IO's

File: src/main/scala/org/constellation/consensus/EdgeProcessor.scala [hunk 5]
```diff
@@ -469,19 +470,14 @@ object Snapshot {
...
-        // To allow consensus more time since the latest snapshot includes all data up to present, but this is simple for now
-        dao.addressService.transferSnapshot(tx)
-          .flatMap(_ => dao.transactionService.memPool.remove(tx.hash))
-//          .flatMap(_ => dao.acceptedTransactionService.remove(Set(tx.hash)))
-          .flatMap(_ => IO { println(s"------- Should remove tx: ${tx.hash}") })
-          .flatTap(_ => dao.metrics.incrementMetricAsync("snapshotAppliedBalance"))
-          .unsafeRunSync()
-      }
+      // DEBUG: mwadon \/
+      val transactions = CheckpointService.fetchTransactions(cb.transactionsMerkleRoot).toList
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/consensus/EdgeProcessor.scala]: Yes, it was done like that and it was causing an error - looks like I'm removing data too fast. Thus I've splitted that into parts to debug it (that's why also a comment)

File: src/main/scala/org/constellation/primitives/storage/TransactionService.scala [hunk 6]
```diff
@@ -1,90 +1,166 @@
...
 
-class TransactionMemPool(size: Int = 50000)
-    extends StorageService[TransactionCacheData](size, Some(240))
+class DefaultTransactionService(dao: DAO) extends TransactionService[String, TransactionCacheData] with StrictLogging {
+  val merklePool = new StorageService[Seq[String]]()
 
-object TransactionService {
-  def apply(implicit dao: DAO, size: Int = 50000) = new TransactionService(dao, size)
-}
+  val pending = new PendingTransactionsMemPool()(dao)
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/primitives/storage/TransactionService.scala]: probably yes, we can rethink it for next PRs

File: src/main/scala/org/constellation/util/Metrics.scala (hunk 4 above)
Comment (marcinwadon) [reply src/main/scala/org/constellation/util/Metrics.scala]: :+1:

File: src/main/scala/org/constellation/p2p/PeerAPI.scala [hunk 7]
```diff
@@ -240,9 +234,13 @@ class PeerAPI(override val ipManager: IPManager, nodeActor: ActorRef)(implicit s
       put {
         entity(as[Transaction]) { tx =>
           dao.metrics.incrementMetric("transactionRXByPeerAPI")
-          dao.transactionService.memPool.updateSync(tx.hash, { tcd =>
-            tcd
-          }, TransactionCacheData(tx))
+
+          dao.transactionService.exists(tx.hash)
```
[the remark this answers is not in the document]
Comment (marcinwadon) [reply src/main/scala/org/constellation/p2p/PeerAPI.scala]: haven't know

**Pull request merged**